opensfm/src/map/src/ba_helpers.cc

```diff
@@ -0,0 +1,528 @@
+      const Vec3d bearing = shot.shot_camera_->Bearing(obs.projection_);
+      const auto& shot_pose = shot.GetPose();
+      bs.row(added) = shot_pose.RotationCameraToWorld() * bearing;
+      os.row(added) = shot_pose.GetOrigin();
+      ++added;
+    }
+  }
+  bs.conservativeResize(added, Eigen::NoChange);
+  os.conservativeResize(added, Eigen::NoChange);
+  if (added >= 2) {
+    const std::vector<double> thresholds(added, reproj_threshold);
+    const auto& res = geometry::TriangulateBearingsMidpoint(os, bs, thresholds,
+                                                            min_ray_angle);
+    coordinates = res.second;
+    return res.first;
+  }
+  return false;
+}
+
+// Add Ground Control Points constraints to the bundle problem
+void BAHelpers::AddGCPToBundle(
+    BundleAdjuster& ba, const AlignedVector<map::GroundControlPoint>& gcp,
+    const std::unordered_map<map::ShotId, map::Shot>& shots) {
+  for (const auto& point : gcp) {
+    const auto point_id = "gcp-" + point.id_;
+    Vec3d coordinates;
+    if (!TriangulateGCP(point, shots, coordinates)) {
+      if (point.coordinates_.HasValue()) {
+        coordinates = point.coordinates_.Value();
+      } else {
+        continue;
+      }
+    }
+    constexpr auto point_constant{false};
+    ba.AddPoint(point_id, coordinates, point_constant);
+    if (point.coordinates_.HasValue()) {
+      const auto point_type = point.has_altitude_ ? PositionConstraintType::XYZ
+                                                  : PositionConstraintType::XY;
+      ba.AddPointPositionWorld(point_id, point.coordinates_.Value(), 0.1,
+                               point_type);
+    }
+
+    // Now iterate through the observations
+    for (const auto& obs : point.observations_) {
+      const auto& shot_id = obs.shot_id_;
+      if (shots.count(shot_id) > 0) {
+        constexpr double scale{0.0001};
+        ba.AddPointProjectionObservation(shot_id, point_id, obs.projection_[0],
+                                         obs.projection_[1], scale);
+      }
+    }
+  }
+}
+
+py::dict BAHelpers::Bundle(
+    map::Map& map,
+    const std::unordered_map<map::CameraId, Camera>& camera_priors,
+    const AlignedVector<map::GroundControlPoint>& gcp, const py::dict& config) {
+  py::dict report;
+
+  auto ba = BundleAdjuster();
+  const bool fix_cameras = !config["optimize_camera_parameters"].cast<bool>();
+  ba.SetUseAnalyticDerivatives(
+      config["bundle_analytic_derivatives"].cast<bool>());
+  const auto start = std::chrono::high_resolution_clock::now();
+
+  for (const auto& cam_pair : map.GetCameras()) {
+    const auto& cam = cam_pair.second;
+    const auto& cam_prior = camera_priors.at(cam.id);
+    ba.AddCamera(cam.id, cam, cam_prior, fix_cameras);
+  }
+
+  for (const auto& pt_pair : map.GetAllLandmarks()) {
+    const auto& pt = pt_pair.second;
+    ba.AddPoint(pt.id_, pt.GetGlobalPos(), false);
+  }
+
+  // Alignemnt method
+  // TODO: Alignment method, auto and vertical
```

Comment:
Couldn't you pass the align method to the C++ method directly ? That way, you just move the L132-135 out of `bundle()` to get the align method, et then you use it straight into the C++ code. That would involve less rewrite.

Reply:
Yes, that would have worked. Now, I already ported the alignment method to C++.